projecttacoma/fqm-testify
# Element reference codes

# Summary
When testing fqm-testify with the "DRCommunicationWithPhysicianManagingDiabetesFHIR-bundle" bundle, Dave noticed that the app crashed and was not able to display data elements with direct reference codes.

## New Behavior
All of the components of the app are the same, but now when a data element doesn't have a valueSet but has a direct reference code, that is what is displayed in the test resource display panel and the scaffolded FHIR resource JSON. In addition, the code and display shown under each test resource for a patient display the direct reference code and display if it exists.

In general, the application now adds all code paths of type FHIR.CodeableConcept, FHIR.Coding, and FHIR.code to the scaffolded FHIR resource JSON. Changes in associated files were made to reflect the changes in parseCodePaths.js.

## Code Changes
- `types.ts` - the structure of the codePaths.ts file was modified so that each dataRequirement object can have more than one path object.
- `parseCodePaths.ts` / `parsePrimaryCodePaths.ts` - modified script that now parses all code type paths for data requirements rather than just the primary code path which is why `parsePrimaryCodePaths.ts` was renamed to `parseCodePaths.ts`. These changes are shown in the previous PR for this with the addition of being able to find elements of type System.string that are embedded in FHIR.Elements. These are just FHIR.code type paths so they are now included in the parse code paths output file.
- `codePaths.ts` / `primaryCodePaths.ts` - `codePaths.ts` replaced `primaryCodePaths.ts` because the new script parses all code paths of all code types into that file. not just the primaryCodePath. The file was renamed to reflect this change.
- `fhir.ts` - functions such as `createFhirResourceString`, `getFhirResourceSummary`, and `getResourcePrimaryCode` were modified to reflect the changes in `codePaths.ts` as well display direct reference codes and add them to the scaffolded Fhir resource JSON.
- `fhir.test.ts` - unit tests updated to reflect changes to `fhir.ts`.
- `package.json` - fixed small mix up.

# Testing Guidance
1. Please ask me questions if any of this doesn't make sense.
2. Run `npm run test` and make sure that all of the tests pass. 
3. Run `npm run dev`
4. Upload the measure bundle "DRCommunicationWithPhysicianManagingDiabetesFHIR-bundle" that is linked on the Jira ticket for this task (#1108). It is also in the slack thread that is linked below. 
5. Create a new patient. Scroll through the available data elements. There are two "Communication" elements that have direct reference codes rather than valuesets. They should look like the following:

<img width="570" alt="image" src="https://user-images.githubusercontent.com/30158156/187484117-b8e250ea-69b8-4b8e-928f-a9ba3a1bba08.png">

6. Try uploading other measure bundles, with or without direct reference codes, and make sure nothing breaks and everything looks correct.
